Verify top-level delegated targets signatures

wellsie1116 opened the pull request:
The previous implementation did not consider delegated roles from the
top-level "targets" role when looking for keys to verify the delegated
targets with, which would result in delegated targets being accepted by
Tuf without signature verification.

Commit:
Verify top-level delegated targets signatures

The previous implementation did not consider delegated roles from the
top-level "targets" role when looking for keys to verify the delegated
targets with, which would result in delegated targets being accepted by
Tuf without signature verification.

erickt commented on src/tuf.rs:
```diff
@@ -438,27 +479,23 @@ impl<D: DataInterchange> Tuf<D> {
-                    Some(d) => d,
-                    None => &targets_delegations,
-                };
+            // FIXME Since the delegating targets defines the valid keys a delegated targets can be
```
Can you file a ticket for this fixme?

wellsie1116 replied:
#279

erickt commented on src/tuf.rs:
```diff
@@ -438,27 +479,23 @@ impl<D: DataInterchange> Tuf<D> {
+            // target path and invalid when accessed via another. Instead of verifying all known
+            // references to this delegation, this verification step should be choosing just the
+            // one for the target path the outer tuf::Client is trying to resolve, but that
+            // currently depends on state in tuf::Client to do so.
```
Can you file a ticket for this fixme?

Ideally `Tuf` wouldn't have to trust outside state from `tuf::Client`. For the short term though, what do you think about `update_delegation` also getting passed in the parent `Role` name? Then we could directly look that up in our map and use that to verify this delegation. That should allow us to support the whole "this delegation is valid when found through this delegation chain, but not valid when found through that delegation chain".

wellsie1116 replied:
That is a good idea for a short term fix.  Implemented.

erickt commented on src/tuf.rs:
```diff
@@ -395,6 +395,50 @@ impl<D: DataInterchange> Tuf<D> {
 
+    /// Walk the base targets delegations and any known nested delegations for the authorized
+    /// signing keys and metadata for the delegation given by `role`.
+    fn find_delegation(&self, role: &MetadataPath) -> Vec<(Vec<&PublicKey>, &Delegation)> {
```
Good old tree searching. Unfortunately delegations could potentially have a loop, so can you add in a "I visited this node already" set to avoid infinite loops?

wellsie1116 replied:
Of course, why wouldn't that be supported :)

Fixed by finding just the delegation we are looking for without having to walk the tree.

erickt reviewed (changes requested):
Finally, we really need some test coverage here. Would you want to add it? Or we could add a "delegation" feature flag to compile out support until it's ready?

Commit:
Provide update_delegations parent role

So it can verify that the provided delegations are valid from the point
of view of the parent. This avoids the need to check more than 1
signature in update_delegations.

Commit:
Add invalid delegations tests

wellsie1116 commented:
There were already some integration style tests in integration.rs, which is how I found this issue in the first place.  I have extended them to:
 * verify signatures of the first level of delegated targets are being verified (and verified that test used to fail), and
 * verify that delegations can be considered valid or invalid, depending on the path to them.

wellsie1116 commented on src/client.rs:
```diff
@@ -909,7 +909,7 @@ where
                             target,
                             snapshot,
-                            Some(meta.as_ref()),
+                            Some((meta.as_ref(), delegation.role().clone())),
```
Am I using the right role here?  I think so, but the integration test also wouldn't catch a mistake here.

erickt replied:
Yeah I think this is the right role.
